File: 11_Graphs/13_Word_Ladder/0127-word-ladder.cpp
/*
Problem: LeetCode 127 - Word Ladder

Description:
Given two words, beginWord and endWord, and a dictionary wordList, return the length of the shortest transformation sequence from beginWord to endWord, such that:
- Only one letter can be changed at a time.
- Each transformed word must exist in the wordList.

If there is no such transformation sequence, return 0.

Intuition:
This problem can be solved using a graph traversal algorithm such as breadth-first search (BFS). We can treat each word as a node in the graph and connect the words that differ by a single character. The problem then reduces to finding the shortest path from the beginWord to the endWord.

Approach:
1. Build a set from the wordList for efficient lookup.
2. Create a queue for BFS traversal and initialize it with the beginWord.
3. Initialize a distance variable to track the length of the transformation sequence.
4. Perform BFS:
   - Pop the front word from the queue.
   - For each character position in the word, replace it with each of the 26 alphabets and check if the resulting word is in the wordList.
   - If the resulting word is the endWord, return the current distance + 1.
   - If the resulting word is in the wordList, add it to the queue and remove it from the wordList to avoid revisiting.
5. If no transformation sequence is found, return 0.

Time Complexity:
The time complexity is O(N * M^2), where N is the number of words in the wordList and M is the length of the words. In the worst case, we may need to explore all possible transformations of each word.

Space Complexity:
The space complexity is O(N * M^2), where N is the number of words in the wordList and M is the length of the words. This is the space used for the wordList set, queue, and visited set.
*/

class Solution {
  public:
    int ladderLength(string beginWord, string endWord, vector<string> &wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end());  // Convert wordList to a set for efficient lookup

        if (wordSet.find(endWord) == wordSet.end()) {
            return 0;  // endWord is not in the wordList, no transformation sequence possible
        }

        queue<string> q;
        q.push(beginWord);
        int distance = 1;

        while (!q.empty()) {
            int levelSize = q.size();

            for (int i = 0; i < levelSize; ++i) {
                string currentWord = q.front();
                q.pop();

                // Check each character position of the word and replace it with all possible alphabets
                for (int j = 0; j < currentWord.length(); ++j) {
                    char originalChar = currentWord[j];

                    for (char c = 'a'; c <= 'z'; ++c) {
                        currentWord[j] = c;

                        if (currentWord == endWord) {
                            return distance + 1;  // Transformation sequence found, return the distance
                        }

                        if (wordSet.find(currentWord) != wordSet.end()) {
                            q.push(currentWord);  // Add the transformed word to the queue
                            wordSet.erase(currentWord);  // Remove the transformed word from the wordSet
                        }
                    }

                    currentWord[j] = originalChar;  // Revert back the character
                }
            }

            ++distance;  // Increment the distance for each level of BFS traversal
        }

        return 0;  // No transformation sequence found
    }
};

// Bidirectional BFS
/*
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end()); // Convert wordList to a set for efficient lookup

        if (wordSet.find(endWord) == wordSet.end()) {
            return 0; // endWord is not in the wordList, no transformation sequence possible
        }

        unordered_set<string> beginSet, endSet; // Sets for the current level of traversal from the beginWord and endWord
        beginSet.insert(beginWord);
        endSet.insert(endWord);

        int distance = 1; // Distance between beginWord and endWord

        while (!beginSet.empty() && !endSet.empty()) {
            // Always expand the smaller set to reduce the search space
            if (beginSet.size() > endSet.size()) {
                swap(beginSet, endSet);
            }

            unordered_set<string> temp; // Temporary set to store the next level of words

            for (const string& word : beginSet) {
                string currentWord = word;

                // Check each character position of the word and replace it with all possible alphabets
                for (int i = 0; i < currentWord.length(); ++i) {
                    char originalChar = currentWord[i];

                    for (char c = 'a'; c <= 'z'; ++c) {
                        currentWord[i] = c;

                        if (endSet.find(currentWord) != endSet.end()) {
                            return distance + 1; // Transformation sequence found, return the distance
                        }

                        if (wordSet.find(currentWord) != wordSet.end()) {
                            temp.insert(currentWord); // Add the transformed word to the next level set
                            wordSet.erase(currentWord); // Remove the transformed word from the wordSet
                        }
                    }

                    currentWord[i] = originalChar; // Revert back the character
                }
            }

            swap(beginSet, temp); // Update the current level set with the next level set
            ++distance; // Increment the distance for each level of BFS traversal
        }

        return 0; // No transformation sequence found
    }
};
*/

File: 11_Graphs/13_Word_Ladder/0127-word-ladder.rs
/*
Problem: LeetCode 127 - Word Ladder

Key Idea:
The key idea is to treat the given word list as a graph, where each word is a node, and two words are connected if they differ by only one character. We can then perform a breadth-first search (BFS) to find the shortest transformation sequence from the beginWord to the endWord.

Approach:
1. Build a set for fast word lookup in the word list.
2. Initialize a queue for BFS, starting with the beginWord.
3. Initialize a set to keep track of visited words.
4. Initialize a level variable to track the depth of the BFS.
5. Perform BFS:
   - For each word in the current level of the queue:
     - Generate all possible one-character transformations of the word.
     - Check if each transformation is in the word list and not visited.
     - If a transformation is the endWord, return the current level + 1 (depth).
     - Otherwise, add the transformation to the queue for the next level.
     - Mark the transformation as visited.
   - Increment the level.
6. If the BFS completes without finding the endWord, return 0 (no transformation sequence found).

Time Complexity:
O(M^2 * N), where M is the length of each word, and N is the total number of words in the word list. Generating all possible transformations for a word takes O(M^2) time, and we do this for each word in the word list.

Space Complexity:
O(N) for the word set and visited set, and O(M) for the queue and transformations.
*/

use std::collections::{HashSet, VecDeque};

impl Solution {
    pub fn ladder_length(begin_word: String, end_word: String, word_list: Vec<String>) -> i32 {
        let word_set: HashSet<String> = word_list.into_iter().collect();
        let mut visited: HashSet<String> = HashSet::new();
        let mut queue: VecDeque<String> = VecDeque::new();
        queue.push_back(begin_word.clone());
        visited.insert(begin_word.clone());
        let mut level = 1;

        while !queue.is_empty() {
            let level_size = queue.len();

            for _ in 0..level_size {
                if let Some(word) = queue.pop_front() {
                    let chars: Vec<char> = word.chars().collect();

                    for i in 0..chars.len() {
                        let original_char = chars[i];

                        for c in b'a'..=b'z' {
                            chars[i] = c as char;
                            let new_word: String = chars.iter().collect();

                            if word_set.contains(&new_word) {
                                if new_word == end_word {
                                    return level + 1;
                                }

                                if !visited.contains(&new_word) {
                                    visited.insert(new_word.clone());
                                    queue.push_back(new_word.clone());
                                }
                            }
                        }

                        chars[i] = original_char;
                    }
                }
            }

            level += 1;
        }

        0
    }
}
